utility/variant_writer_compat.cpp

```diff
@@ -0,0 +1,471 @@
+#include "variant_writer_compat.h"
+#include "core/input/input_event.h"
+#include "core/io/resource_loader.h"
+#include "core/os/keyboard.h"
+#include "core/string/string_buffer.h"
+#include "core/io/image.h"
+#include "core/variant/variant_parser.h"
+#include "image_parser_v2.h"
+#include "resource_loader_compat.h"
+
+namespace ToV4{
+
+enum Type {
+
+    NIL,
+    // atomic types
+    BOOL,
+    INT,
+    REAL,
+    STRING,
+    // math types
+	VECTOR2, 
+	VECTOR2I, //Unused
+	RECT2,
+	RECT2I,   //Unused 
+	VECTOR3, 
+	VECTOR3I,  //Unused
+	TRANSFORM2D,
+    PLANE,
+    QUAT, // 10
+    _AABB, //sorry naming convention fail :( not like it's used often
+    BASIS,  //Basis
+    TRANSFORM,
+    // misc types
+    COLOR,
+    IMAGE, // 15
+    NODE_PATH,
+    _RID,
+    OBJECT,
+    INPUT_EVENT, // v2 struct which isn't mapped
+	SIGNAL, //Unused
+    DICTIONARY, // 20
+    ARRAY,
+	// V3 arrays
+	POOL_BYTE_ARRAY = 26, // 20
+	POOL_INT_ARRAY = 27,
+	POOL_INT64_ARRAY = 28, //Unused
+	POOL_REAL_ARRAY = 29,
+	POOL_REAL64_ARRAY = 30, //Unused
+	POOL_STRING_ARRAY = 31,
+	POOL_VECTOR2_ARRAY = 32,
+	POOL_VECTOR3_ARRAY = 33, // 25
+	POOL_COLOR_ARRAY = 34,
+    VARIANT_MAX = 35,
+    // V2 arrays
+    RAW_ARRAY = 26,//Byte array
+    INT_ARRAY = 27, 
+	INT64_ARRAY = 28, //unused
+    REAL_ARRAY = 29,
+	REAL64_ARRAY = 30, //Unused
+    STRING_ARRAY = 31, // 25
+    VECTOR2_ARRAY = 32,
+    VECTOR3_ARRAY = 33,
+    COLOR_ARRAY = 34,
+	// v2 name for transform2d
+	MATRIX32 = 11,
+	// v2 name for matrix3
+	MATRIX3 = 15,
+};
+}
+//////////////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+static String rtosfix(double p_value) {
+
+	if (p_value == 0.0)
+		return "0"; //avoid negative zero (-0) being written, which may annoy git, svn, etc. for changes when they don't exist.
+	else
+		return rtoss(p_value);
+}
+
+Error VariantWriterCompat::write_compat( const Variant &p_variant, const uint32_t ver_major, StoreStringFunc p_store_string_func, void *p_store_string_ud, EncodeResourceFunc p_encode_res_func, void *p_encode_res_ud){
+	
+	// use the v4 write function instead for v4
+	if (ver_major == 4){
+		return VariantWriter::write(p_variant, p_store_string_func, p_encode_res_ud, p_encode_res_func, p_encode_res_ud);
+	}
+	
+	switch (p_variant.get_type()) {
```

Comment:
```suggestion
	switch ((ToV4::Type)p_variant.get_type()) {
```
Cast to suppres `comparison of different enumeration types in switch statement` warnings.